### PR TITLE
search: separate parameters and pattern types in parse tree

### PR DESCRIPTION
**Pull request by rvantonder**
TL;DR: 

The change separates a singular datatype `Parameter` that currently represents both search patterns and `field:value` parameters to be separate (a new `Pattern` type is created). Everything else is updated around that change (separate functions for parsing these, visitor interface, tree reduction, etc.).

We can have confidence in the change because tests pass, and affected tests are updated to the correct desired behavior.

---

Long description so I can trace my thinking and how this changed.

The thing motivating this change is that we currently use a convention that search patterns are represented as a `type Parameter` in the AST where field is `""` and value represents the pattern value. This same convention is in the existing (older) parser and can simplify some implementation details. Initially I didn't think we need to separate `field:value` versus patterns (`"":value`) at the parsing level, but since https://github.com/sourcegraph/sourcegraph/pull/9844 combining the two in the same datatype started bothering me because we do have some properties that we semantically distinguish in patterns versus parameters:

> Adding Quoted is not the desired end goal for the Parameter types. It currently pollutes the visitor interface a bit. I will be turning parameter values into better structures with type members later.

Basically, we want to track whether a search pattern was quoted (it has semantic implications for our search: quoting in regex mode implies literal search) but we do not care about that for `field:value` parameters, since we will process `field:value` versus `field:"value"` or `field:'value'` all the same.

Thus, while this change adds extra logic to deal with the extra `Pattern` case, it produces a final tree that better introspection, simplifies the visitor interface, and simplifies other parts of query validation.

### CODE REVIEW

File: internal/search/query/mapper.go [hunk 1]
```diff
@@ -37,6 +40,11 @@ func (*BaseMapper) MapParameter(visitor Mapper, field, value string, negated boo
 	return Parameter{Field: field, Value: value, Negated: negated}
 }
 
+// Base mapper for Patterns. It is the identity function.
+func (*BaseMapper) MapPattern(visitor Mapper, value string, negated bool, quoted bool) Node {
```
Comment (unknwon) [inline internal/search/query/mapper.go:44]: According to linter:
```suggestion
func (*BaseMapper) MapPattern(visitor Mapper, value string, negated, quoted bool) Node {
```

File: internal/search/query/parser.go [hunk 2]
```diff
@@ -562,72 +579,86 @@ func (p *parser) TryParseDelimiter() (string, bool) {
...
 
-// ParseValue parses a value at the current position and whether it was quoted.
-// A value may belong to a parameter like repo:<value> or the <value> may be a
-// search pattern. Values may be quoted. ParseValue cannot fail: it will first
-// try to scan well-formed delimiters, like quoted strings. If that fails, it
-// will accept unbalanced strings as patterns. Uses of value can then be
-// validated along other concerns for different search types (literal, regexp,
-// etc.).
-func (p *parser) ParseValue() (string, bool) {
+// ParseFieldValue parses a value followed by a field like "repo:". If the value
```
Comment (unknwon) [inline internal/search/query/parser.go:582]: "a value followed by a field" sounds like "repos:" appears _after_ the value? Which I think is not true (i.e. the opposite)?

File: internal/search/query/types.go [hunk 3]
```diff
@@ -102,16 +102,27 @@ func (q AndOrQuery) StringValue(field string) (value, negatedValue string) {
...
 }
 
 func (q AndOrQuery) Fields() map[string][]*types.Value {
 	fields := make(map[string][]*types.Value)
-	VisitParameter(q.Query, func(field, value string, _, quoted bool) {
-		fields[field] = valueToTypedValue(field, value, quoted)
+	VisitPattern(q.Query, func(value string, _, quoted bool) {
+		// FIXME. NOTE: overrides. Seems not good, how is Fields() used?
+		// For existence? What about "content"?
+		fields[""] = valueToTypedValue("", value, quoted)
```
Comment (unknwon) [inline internal/search/query/types.go:122]: Does this look right? (definitely should avoid giving meaning to empty string whenever possible)
```suggestion
		fields["content"] = valueToTypedValue("", value, quoted)
```

Or is it possible `fields["content"]` previously have value that we want to concatenate?

Comment (rvantonder) [reply internal/search/query/types.go]: Ah, thanks for catching this comment, I meant to address and remove it.

The issue here is that `Fields()` satisfies an existing interface of "values that our search code uses to figure out what to do". Here are [all the uses](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+%5C.Fields%5C%28%5C%29+lang:go&patternType=regexp#1). It's not clear how, what, or whether something is expected for `Fields()[FieldDefault]` AKA `Fields()[""]`. The empty string here is exactly how we currently store this information in `Fields()`, so I am just doing the same and trying to satisfy the interface (though, thanks to this PR, that will not be needed once we migrate away). After looking at everything, the safe thing to do is right now is to rely on `Values()` to supply the `Value` of fields, despite this doing extra work for visiting the query. I'll remove this interface later.
